proposals/exception-handling/Exceptions-formal-overview.md

```diff
@@ -0,0 +1,211 @@
+
+## Abstract Syntax
+
+### Types
+
+Tag Types
+
+```
+tagtype ::= [t*]→[]
+```
+
+### Instructions
+
+```
+instr ::= ... | throw x | rethrow l
+        | try bt instr* (catch x instr*)* (catch_all instr*)? end
+        | try bt instr* delegate l
+```
+
+### Modules
+
+Tags
+
+```
+tag ::= export* tag exntype  | export* tag tagtype import
+```
+
+Modules
+
+```
+mod ::= module ... tag*
+```
+
+## Validation (Typing)
+
+#### Modification to labels
+
+To verify that a `try...delegate l` instruction refers to a label surrounding the instructions of a try block (call this a try-label), introduce a `kind` attribute to labels in the validation context, which is set to `try` when the label is a try-label.
+
+Similarly, to verify that the `rethrow l` instruction refers to a label surrounding the instructions of a catch block (call this a catch-label), we allow the `kind` attribute of labels in the validation context to be set to `catch` when the label is a catch-label.
+
+The original notation `labels [t*]` is now an abbreviation for:
+
+```
+labels {result [t*], kind ε}
+```
+
+
+### Instructions
+
+
+```
+C.tags[x] = [t*]→[]
+-----------------------------
+C ⊢ throw x : [t1* t*]→[t2*]
+
+
+C.labels[l].kind = catch
+----------------------------
+C ⊢ rethrow l : [t1*]→[t2*]
+
+
+C.types[bt] = [t1*]→[t2*]
+C, labels {result [t2*], kind try} ⊢ instr* : [t1*]→[t2*]
+(C.tags[x] = [t*]→[] ∧
+ C, labels { result [t2*], kind catch } ⊢ instr* : [t*]→[t2*])*
+(C, labels { result [t2*], kind catch } ⊢ instr'* : []→[t2*])?
+-----------------------------------------------------------------------------
+C ⊢ try bt instr* (catch x instr'*)* (catch_all instr''*)? end : 	[t1*]→[t2*]
+
+
+C.types[bt] = [t1*]→[t2*]
+C, labels {result [t2*], kind try} ⊢ instr* : [t1*]→[t2*]
+C.labels[l].kind = try
+----------------------------------------------------------
+C ⊢ try bt instr* delegate l : [t1*]→[t2*]
+```
+
+## Execution (Reduction)
+
+### Runtime structure
+
+Stores
+
+```
+S ::= {..., tags exninst*}
+```
+
+Tag Instances
+
+```
+taginst ::= {type tagtype}
+```
+
+Module Instances
+
+```
+m ::= {..., tags a*}
+```
+
+Administrative Instructions
+
+```
+instr ::= ... | throw a | catch_n{a? instr*}* instr* end
+        | delegate{l} instr* end | caught_m{a val^n} instr* end
+```
+
+Block contexts and label kinds
+
+So far block contexts are only used in the reduction of `br l` and `return`, and only include labels or values on the stack above the hole. If we want to be able to break jumping over try-catch and try-delegate blocks, we must allow for the new administrative control instructions to appear after labels in block contexts, mirroring the label kinds of labels in validation contexts.
+
+```
+label_kind ::= try_label_kind | catch_label_kind
+try_label_kind ::= catch_m{a? instr*}* | delegate{l}
+catch_label_kind ::= caught_m{a val*}
+B^0 ::= val* [_] instr*
+B^{k+1} ::= val* label_n{instr*} label_kind? B^k end? end instr*
```

Comment:
This seems a bit off, since label_kind and end would be optional independently from each other.
```suggestion
B^k ::= val* B'^k instr*
B'^0 ::= [_]
B'^{k+1} ::= label_n{instr*} B^k end | catch_m{a? instr*}* B^{k+1} end | caught_m{a val*} B^{k+1} end | delegate{l} B^{k+1} end
```